### PR TITLE
Attempt to parse comments immediately after colon

### LINKED ISSUE #5603: SCSS single-line comments are formatted incorrectly
**Important Note:**
Prettier has nice support for Grid-Template-Areas . See [here](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEA6A5gJwJYBMAEwAOlPvlngLTwC2ADgDYCG8lTmcTAzkiWWURAALTrjiYyIpmIn4pMwX36CueOACN2+SLAQxt0eLEWllIAGYQI8WZeviydm4IDcJAL4gANCAh0Y2NBcyKDsmBAA7gAK7AjBKEwAbhB43iDqmExgANZwMADKdFnYUOjIMJgArnA+QjA0DADqQtjwXEVgcPlxrdiJrQCeyOBcwT4lXOIwUZnoNEzI5kwMkz4AVlwAHgBCmTl5+Uw0cAAyJXCLy6sgG5v5JegMcACKlfaXKzUgRZiTmMNcMCjNJ0HCwRp4GBCZAADgADD5QRBJo1MnRhqC4H9EhcfBwAI6VbAcGZMOYLJBLT4+SY0bDlKpfVSlJ6vd6Uq5fGBMdQQ3BQ5AAFh8FSY2AYDwAwhAaPNhlBoLiQJVJgAVHnxKnXbgwD7XXAQMB6r4QSowOhmgBMxvc7iAA )

The example shows only how to reproduce the issue with single-line comments!

**Prettier 1.15.3**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAKgCwJYDOGhGAygMKmkB0GACgE5wwx5wMYC2EAbnETBxwMAQyYiMBAA5wweAGZ4wI1tAwBXKRgBGcfFAAmGQcPQZInTghgAdKNQDmDPEeB2MHzx6cuAtPE4pABsVOF8xOBECJAx0dy8EmxAhEQN2DBS0jkz2JNjUeITPJIIXOG0xc2h4WCrYazy4qCKvJPkICHgOds70nq6kgG53AF8QABoQCClVKAJkUDEGCAB3Ogi55BARHggXCZBtBhEwAGtmUikTvCgHZBgGdThJnBhOIIB1fHhpE7hSBClVg8PAwACeWzABHmkxuBHYMEYIgcnBEyHkIiC8MmACsCAAPABCxzOFxEVgAMjc4OjMdiQHj8aQbg4gnAAIrqXq0rHPEBXBjwhhbAhQmH85ywD4uQTIAAcAAZJlJlvCPscpFsVfx2HwDkwAI7qPBMJEotFIDG8ybwzh4e6PPmlW5szncy10vkwETaaUGWVIABMkweIjwQRZ5AglgtICg0Bpk3U8KwPvmHt5IxGQA)
```sh
--parser scss
```

**Input:**
```scss
// This is SCSS. Prettier moves the area specification up behind the // comment
.grid {
        grid-template-areas: //
          "header header header" //
          "sidebar content content" //
          "footer footer footer";
 }
```

**Output:**
```scss
// This is SCSS. Prettier moves the area specification up behind the // comment
.grid {
  grid-template-areas: // "header header header"//
    "sidebar content content"//
    "footer footer footer";
}
```

**Expected behavior:**
```scss
// This is SCSS. Prettier moves the area specification up behind the // comment
.grid {
  grid-template-areas: // 
    "header header header" //
    "sidebar content content" //
    "footer footer footer";
}
```

There are related (closed) Issues for older Versions:
- https://github.com/prettier/prettier/issues/4311
- https://github.com/prettier/prettier/issues/4594

### PR DESCRIPTION
**Pull request by robert-j-webb**
Fixes #5603 
I am not certain this is the best approach but this definitely fixes the issue.

- [X] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [X] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fisker** [comment]
Can we  add this to   https://github.com/prettier/prettier/blob/5bd45d79afe13c9593d53e03903e0c04f0c463c9/src/language-css/parser-postcss.js#L183

Like  https://github.com/prettier/prettier/blob/5bd45d79afe13c9593d53e03903e0c04f0c463c9/src/language-css/parser-postcss.js#L203-L212

**Comment by nex3** [comment]
Would fixing https://github.com/shellscape/postcss-values-parser/issues/105 help with this?

**Comment by robert-j-webb** [comment]
@nex3 - It looks like postcss-value-parser has no support for single line comments. This is the only comment parsing that they're doing: https://github.com/TrySound/postcss-value-parser/blob/master/lib/parse.js#L103

@fisker 
I agree that the correct answer is to add some sort of parsing that catches this situation. I'll see how it goes if I add something to that function.

**Comment by robert-j-webb** [comment]
@fisker 

I did some debugging and the value at that point in parseValue is 

```
 "header header header" // b
        "sidebar content content" // c
        "footer footer footer"
```

With an input of:
```
 grid-template-areas: // a
    "header header header" // b
    "sidebar content content" // c
    "footer footer footer";
```

So the issue isn't so much that postcss-values-parser is having a problem but rather that something upstream is over parsing.

**Comment by robert-j-webb** [comment]
So I switched from modifying printer to modifying parser, and I went with this approach. This is inside `parseNestedCss`, but I could move it to somewhere else.

```
 // Move comments that are stuck in between into value.
    if (node.raws.between && node.raws.between.match(/\/\/|\/\*/)) {
      const commentIdx = node.raws.between.search(/\/\/|\/\*/);
      if (commentIdx !== -1) {
        const beforeComment = node.raws.between.slice(0, commentIdx);
        const afterComment = node.raws.between.slice(commentIdx);
        node.raws.between = beforeComment;
        node.raws.value.scss = afterComment + node.raws.value.scss;
        node.raws.value.raw = afterComment + node.raws.value.raw;
      }
    }
```

Basically this fixes the between parsing failing by manually doing it. Let me know if that works.

**Comment by robert-j-webb** [comment]
I found some issues when it was parsing CSS and I couldn't make heads or tails of how the node was different in the CSS context so I made this particular fix SCSS specific.

**Comment by thorn0** [comment]
Is this related or a separate bug?

**Prettier pr-7593**
[Playground link](https://deploy-preview-7593--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEA6A5gJwJYBMAEwAOlPvlngLTwC2ADgDYCG8lTmcTAzkvgPR98TEmTJEQAC0644mfFKYy5CpeP6CARiNHiueOBvb5IsBDGPR4sNQKH8AVBvt9tYkADMIEeHM-fZ+H4+4gDcJAC+JCAANCAQdDDY0FzIoOyYEADuAArsCCkoTABuEHgxIBqYTGAA1nAwAMp01dhQ6MgwmACucLESMDQMAOoS2PBczWBwDflj2EVjAJ7I4Fwpsa1csjDZVeg0TMjuTAxbsQBWXAAeAEJVtfUNTDRwADKtcEcnZyCXVw2tdAMOAARS6-i+p16IGamC2mBWXDAa3KdBwsCGeBgEmQAA4AAyxNEQLZDKp0FZouDwoqfWIAR3B8F28QKIG4lCgcDgMlw5Q4jOwHF2TH2hyQxyhsS2NGwHW60L0bWBYIhEu+0JgTA0mNw2OQACZYp0mNgGICAMIQGgHFZQaB0kBdLYAFW1BUlPyKPQAklAZLAGmAcAkAIL+howRbAyFbcLhIA)
```sh
--parser scss
```

**Input:**
```scss
.grid {
  grid-template-areas: // a
    "header header header" // b
    "sidebar content content" // a /*b*/
    "footer footer footer";
}

```

**Output:**
```scss
.grid {
  grid-template-areas: // a
    "header header header" // b
    "sidebar content content" // a b
    "footer footer footer";
}

```

**Comment by robert-j-webb** [comment]
@thorn0 - This should fix #5603

File: src/language-css/parser-postcss.js [hunk 1]
```diff
@@ -298,6 +298,23 @@ function parseNestedCSS(node, options) {
       node.raws.value = selector;
     }
 
+    // Move comments that are stuck in between into value for sass only
+    if (
+      node.raws.value &&
+      node.raws.value.scss &&
```
Comment (evilebottnawi) [inline src/language-css/parser-postcss.js]: `isScssParser()`

File: src/language-css/parser-postcss.js [hunk 2]
```diff
@@ -298,6 +298,23 @@ function parseNestedCSS(node, options) {
...
+      node.raws.between &&
+      node.raws.between.match(/\/\/|\/\*/)
```
Comment (evilebottnawi) [inline src/language-css/parser-postcss.js]: `hasInlineComment`

Comment (robert-j-webb) [reply src/language-css/parser-postcss.js]: Hmm I can't seem to find that function in the codebase?

Comment (evilebottnawi) [reply src/language-css/parser-postcss.js]: Yes, please move it to utils and rename the function

File: src/language-css/parser-postcss.js (hunk 1 above)
Comment (evilebottnawi) [reply src/language-css/parser-postcss.js]: Please move it to utils and rename the function

Comment (robert-j-webb) [reply src/language-css/parser-postcss.js]: I named it isSCSSParserForNode, since there is already an isSCSSParser function

**Comment by robert-j-webb** [comment]
It is working fine with Less, but it's failing with the CSS parser

[playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEASAtgTwLQBsCWAzjAMxIAEA9JeQEICGAXgDpTnkDkAZhBBwDRUaAMV6t2HAEb0AThwDcQurJD8QEAA4x80QslCyZEAO4AFWQj0p6ANwj4AJqpCSZ9MAGs4MAMob3+FAA5sgwMgCucGoAFjDouADq0fjwhP5gcD6WKfg2KZjI4IR6aoGEcDIwpm5B6PTIXPS45WoAVoQAHrRunt4+9OhwADKBcA1NLSDtHT6BQbhwAIrhEPDjzVEg-jLlMoVgxc4aMoEwCY4w0cgAHAAMascQ5QluGoXHcLs2Y2oAjivwaqaKwgeiEbBQOBwBzQ5wyOD-fDw6r0Wr1JCNDZqcrofChCKbQhzBbLVZjDETTYweiSc4OS7IABMajC9HwBGCAGEIOg6oUoNAfiBwuUACo0qyYyY2SIASSgMNgPjAJy0AEEFT4YJgFutygBffVAA)

```
input
$my-list3: // Baz
  'foo', // Foo
  'bar'; // Bar
```
```
output
$my-list3: // Baz "foo",
  // Foo
    "bar"; // Bar
```

**Comment by robert-j-webb** [comment]
It looks like my code is changing the AST

```
    - Expected
    + Received

                      "groups": Array [
                        Object {
                          "quoted": true,
                          "type": "value-string",
    +                     "value": "header header header",
    +                   },
    +                   Object {
    +                     "inline": true,
    +                     "type": "value-comment",
    +                     "value": " b",
    +                   },
    +                   Object {
    +                     "quoted": true,
    +                     "type": "value-string",
                          "value": "sidebar content content",
                        },
                        Object {
                          "inline": true,
                          "type": "value-comment",
```

Which makes sense because

```
.grid {
  grid-template-areas: // a "header header header" // b
    "sidebar content content" // c
    "footer footer footer";
}
```
becomes: 
```
.grid {
  grid-template-areas: // a
    "header header header" // b
    "sidebar content content" // c
    "footer footer footer";
}
```

Is there any way to make an exemption to the AST checking? Or how else should I deal with this?

**Comment by glen-84** [comment]
Should this be closed now that #5603 was fixed?

**Comment by robert-j-webb** [comment]
Sure, thanks Glen.

**Comment by fisker** [comment]
I missed this PR when I send https://github.com/prettier/prettier/pull/8366, the comments between prop and value still not handled well, so you can improve it, if you want.

**Comment by robert-j-webb** [comment]
> I missed this PR when I send #8366, the comments between prop and value still not handled well, so you can improve it, if you want.

Yeah, I'd like to get this submitted. I see this error come up in about ~0.1% of files.  

I ran into an issue while testing this where changing the parsing inbetween the colon and the first value actually changes the AST, so I don't know how I would merge this without breaking the AST.

**Comment by fisker** [comment]
Please rebase first.

**Comment by robert-j-webb** [comment]
Ok I rebased and it looks like everything is working as intended. We can close out the issue. I left the tests in just to confirm, but this doesn't need to be merged [Unless you want more tests?]

**Comment by fisker** [comment]
Those cases already tested, I thought this might fix those tests https://github.com/prettier/prettier/blob/d7340db7e020be6372aff1d022e4130504f3689a/tests/scss/comments/between-decl.scss#L44-L69 .

Seems not.

**Comment by fisker** [comment]
Thanks anyway.

**Comment by fisker** [comment]
I'm closing this, the original issue is fixed, and the new tests added in this PR already exists.

**Pull request closed without merging**